pkg/distributor/distributor.go

```diff
@@ -2771,8 +2771,15 @@ func (d *Distributor) MetricsForLabelMatchers(ctx context.Context, from, through
 
 	queryLimiter := mimir_limiter.QueryLimiterFromContextWithFallback(ctx)
 
-	result := make([]labels.Labels, 0, len(metrics))
+	resultCap := len(metrics)
+	if hints != nil && hints.Limit > 0 {
+		resultCap = min(resultCap, hints.Limit)
+	}
+	result := make([]labels.Labels, 0, resultCap)
 	for _, m := range metrics {
+		if len(result) >= resultCap {
```

Comment:
Why not limit few lines above when we populate `metrics` to deduplicate series? We could just stop once `len(metrics)` reached the limit.

Similarly, I'm wondering why we don't apply the limiter in the same `for` loop above when we populate `metrics`. Do you see a good reason to not do it there?

Reply:
> Do you see a good reason to not do it there?

I think, the only one is that, in the worst case, this will check every duplicate against the series limiter, calculating the hash of the series twice. I cannot tell if that better or worse, from applying this limit later.